src/searching/saddleback_search.rs

```diff
@@ -1,85 +1,147 @@
+            $(
+                #[test]
+                fn $name() {
+                    let (matrix, element, expected) = $tc;
+                    if let Some(expected_pos) = expected {
+                        assert_eq!(matrix[expected_pos.0][expected_pos.1], element);
+                    }
+                    assert_eq!(saddleback_search(&matrix, element), expected);
+                }
+            )*
+        };
     }
 
-    // Additional test case: Element in the middle of the matrix
-    #[test]
-    fn test_element_in_middle() {
-        let matrix = vec![
-            vec![1, 10, 100, 1000],
-            vec![2, 20, 200, 2000],
-            vec![3, 30, 300, 3000],
-        ];
-        assert_eq!(saddleback_search(&matrix, 200), (2, 3));
+    saddleback_tests! {
+        test_element_not_found: (
+            vec![
+                vec![1, 10, 100],
+                vec![2, 20, 200],
+                vec![3, 30, 300]
+            ],
+            123,
+            None::<(usize, usize)>,
+        ),
+        test_element_at_top_left: (
+            vec![
+                vec![1, 10, 100],
+                vec![2, 20, 200],
+                vec![3, 30, 300]
+            ],
+            1,
+            Some((0, 0)),
+        ),
+        test_element_at_bottom_right: (
+            vec![
+                vec![1, 10, 100],
+                vec![2, 20, 200],
+                vec![3, 30, 300]
+            ],
+            300,
+            Some((2, 2)),
+        ),
+        test_element_at_top_right: (
+            vec![
+                vec![1, 10, 100],
+                vec![2, 20, 200],
+                vec![3, 30, 300]
+            ],
+            100,
+            Some((0, 2)),
+        ),
+        test_element_at_bottom_left: (
+            vec![
+                vec![1, 10, 100],
+                vec![2, 20, 200],
+                vec![3, 30, 300]
+            ],
+            3,
+            Some((2, 0)),
+        ),
+        test_element_in_middle: (
+            vec![
+                vec![1, 10, 100, 1000],
+                vec![2, 20, 200, 2000],
+                vec![3, 30, 300, 3000],
+            ],
+            200,
+            Some((1, 2)),
+        ),
+        test_element_smaller_than_min: (
+            vec![
+                vec![1, 10, 100],
+                vec![2, 20, 200],
+                vec![3, 30, 300],
+            ],
+            0,
+            None::<(usize, usize)>,
+        ),
+        test_horizontal: (
+            vec![
+                vec![1, 10, 100],
+            ],
+            100,
+            Some((0, 2)),
+        ),
+        test_vertical: (
+            vec![
+                vec![1],
+                vec![2],
+                vec![3],
+            ],
+            2,
+            Some((1, 0)),
+        ),
+        test_single_element: (
+            vec![
+                vec![1],
+            ],
+            1,
+            Some((0, 0)),
+        ),
```

Comment:
Such test case fails:
```suggestion
        ),
        test_empty_rows: (
            vec![
                vec![],
                vec![],
                vec![],
                vec![],
            ],
            1,
            None::<(usize, usize)>,
        ),
```

Reply:
I see that, the current implementation does not handle this case. In cases of an empty matrix, or matrix with all empty rows, returning `None` is the best option.